Fix for Issue #397

msat59 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
Early unwanted stopping if `ftol` is used, no matter what threshold is set, because of no change in `best_cost` in two consecutive iterations.


## Related Issue
* [#397](https://github.com/ljvmiranda921/pyswarms/issues/397)

* The logger was also improved to show the termination condition, and a bug in showing the progress bar was fixed too.

* Functions docstrings were also updated.


## Motivation and Context
It avoids having an unwanted stop.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
All GlobalBestPSO, LocalBestPSO, and GeneralOptimizerPSO were tested for Sphere(), with following criteria:
- `n_particles=50, dimensions=2, iters=300`
- without `ftol`  («iters=300» reached)
- `ftol=1e-60`   («iters=300» reached)
- `ftol=1e-3`  («ftol=0.001» reached)
- `ftol=1e-10, ftol_iter=5`  («ftol_iter=5» reached)
- `ftol=1e-60, ftol_iter=50`  («iters=300» reached)

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

msat59 commented:
What's that conflict for? One blank line?

ljvmiranda921 commented:
Hi @msat59 , yeah it seems like it's a blank line, that's weird...
Let me check this PR during the weekend! I hope it's fine with you, work got busy lately.
Let me pull-in @nishnash54 and @hakan-gecili too to give their reviews so that we have an extra pair of eyes! Thanks for this PR!

whzup commented:
Aren't there commits from earlier PRs (verbose, close pool) in here? @ljvmiranda921 @msat59

msat59 commented:
> Aren't there commits from earlier PRs (verbose, close pool) in here? @ljvmiranda921 @msat59

Yes. I don't know why they've appeared. I created PR via website, compared with ljvmiranda921 branch. Did I do something wrong?

nishnash54 commented:
You should have updated your local repository by pulling the `upstream master` and **create a new branch for development**. Commits from your old PR were squashed in the merge so your local log showed additional commits.

You GitHub repo has a message showing exactly this.
> This branch is 20 commits ahead, 2 commits behind ljvmiranda921:master. 

I'm not exactly sure how to fix this issue. Maybe you need to merge `upstream master`.
Any inputs on how to fix this issue? @ljvmiranda921 @whzup

whzup commented:
I also had this problem but that was years ago, can't remember how I solved it 😄. You surely have to [sync your fork](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) first (just do this whenever you start working on another issue). It would surely be good practice to name the branch after whatever you are doing with it (e.g. `fix-ftol`) but I think this might be a bit nasty to do now. Other than that, @ljvmiranda921 knows more about `git` than I do 😄

msat59 commented:
> I also had this problem but that was years ago, can't remember how I solved it . You surely have to [sync your fork](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork) first (just do this whenever you start working on another issue). It would surely be good practice to name the branch after whatever you are doing with it (e.g. `fix-ftol`) but I think this might be a bit nasty to do now. Other than that, @ljvmiranda921 knows more about `git` than I do

I'm using Web UI. I couldn't find a way to exclude the previous commits

ljvmiranda921 commented:
Wait sorry, I'll try if I can do the git operation haha this weekend!

nishnash54 commented:
@msat59 I had some time today to look into this PR.

I was able to get the exact same results to the tests you have mentioned in the description on the current version of the library `pyswarms: master`. I was not sure of what you mean by `«ftol=0.001» reached`, but that is a new implementation in this PR (an exit when there is no improvement over subsequent iterations).

I suggest you can create a sample test bed that describes where the current implementation fails or how you would like to have some different results.

> Early unwanted stopping if `ftol` is used, no matter what threshold is set, because of no change in `best_cost` in two consecutive iterations.

This is the main purpose of `ftol` and if you don't want **two consecutive iterations**, you can specify `ftol_iter` to extend the property over more iterations.

msat59 commented on pyswarms/discrete/binary.py:
```diff
@@ -161,10 +168,16 @@ def optimize(self, objective_func, iters, n_processes=None, **kwargs):
             swarm.
         """
+        # Apply verbosity
+        if verbose:
+            logginglevel = logging.NOTSET
+        else:
+            logginglevel = logging.INFO
```
Yeah, that was my bad. I always had problem with the meaning of verbose. So the condition must be reversed.

ljvmiranda921 replied:
Hi @nishnash54 and @msat59 , yeah we don't have a test for verbose behavior and it's intended: it seems to be more of a big engineering effort to check stdout streams to ensure that verbose is working. Feel free to tell me if you have a different opinion!

msat59 commented:
> @msat59 I had some time today to look into this PR.
> 
> I was able to get the exact same results to the tests you have mentioned in the description on the current version of the library `pyswarms: master`. I was not sure of what you mean by `«ftol=0.001» reached`, but that is a new implementation in this PR (an exit when there is no improvement over subsequent iterations).
> 
> I suggest you can create a sample test bed that describes where the current implementation fails or how you would like to have some different results.
> 
> > Early unwanted stopping if `ftol` is used, no matter what threshold is set, because of no change in `best_cost` in two consecutive iterations.
> 
> This is the main purpose of `ftol` and if you don't want **two consecutive iterations**, you can specify `ftol_iter` to extend the property over more iterations.

@nishnash54, I really cannot explain again what the issue was. I didn't touch the ftol_iter implementation, except skipping it if the ftol_iter==0.

Before implementation of the `ftol_iter` and even now, the code had a bug. That's it.

sample:

> import numpy as np
> import pyswarms as ps
> from pyswarms.utils.functions import single_obj as fx
> 
> options = {'c1': 0.5, 'c2': 0.3, 'w':0.9}
> 
> ndim=2
> max_bound = 5 * np.ones(ndim)
> min_bound = - max_bound
> bounds = (min_bound, max_bound)
> 
> optimizer = ps.single.GlobalBestPSO(n_particles=50, dimensions=ndim, options=options,
>                                     bounds=bounds,
>                                     ftol=1e-60,
>                                     ftol_iter=0
>                                     )
> cost, pos = optimizer.optimize(fx.sphere, iters=300, n_processes=None)
> 
> print(f"Position:{pos}, Cost:{cost}, Stopped at iteration: ", len(optimizer.cost_history))

Output with the old code (without fix): 
```
pyswarms.single.global_best - INFO - Optimization finished | «ftol=1e-60» reached | Last iteration: 3) |
best cost: 0.1340366869131137, best pos: [ 0.19796776 -0.30796989]
```
### which is completely wrong
`ftol_iter` is zero. I want the gradient in changes of in best_cost be less than `1e-60`; but it becomes zero because of the nature of PSO. That's a problem. It is not what I want by setting `ftol=1e-60`.



Output with the fix:
```
pyswarms.single.global_best - INFO - Optimization finished | «iters=300» reached | Last iteration: 300) |
best cost: 1.6088536411402293e-16, best pos: [-2.52131617e-09  1.24309424e-08]
```

### I haven't changed the behavior of `ftol_iter`.  
If you change `ftol_iter==1` in the code provided, the search stops on 2nd iteration. That's what the `ftol_iter` has been designed for.

msat59 commented:
> @msat59 I had some time today to look into this PR.
> 
> I was able to get the exact same results to the tests you have mentioned in the description on the current version of the library `pyswarms: master`. I was not sure of what you mean by `«ftol=0.001» reached`, but that is a new implementation in this PR (an exit when there is no improvement over subsequent iterations).
> 
> I suggest you can create a sample test bed that describes where the current implementation fails or how you would like to have some different results.
> 

Oh...I deleted my response.

@nishnash54 , I have explained the issue for more than 10 times. Bro... I have been using GA (Genetic Algorithm) since 2007 and PSO since 2014 for optimization and feature selection. So trust me. I know how PSO behave and what is wrong here.

The code before and after implemetation of the `ftol_iter` has had a bug.

When I set `ftol=10e-60`, I mean the gradient in changes of the best_cost must be less than `1e-60`, which is a very small value. It literally means never stop the search until max iter reaches.

Because of the bug, the search is stopped when the best_cose doesn't change in two consecutive iterations.**EVEN IF THE ftol_iter=0**.

This is the sample code

> import numpy as np
> import pyswarms as ps
> from pyswarms.utils.functions import single_obj as fx
> 
> options = {'c1': 0.5, 'c2': 0.3, 'w':0.9}
> 
> ndim=2
> max_bound = 5 * np.ones(ndim)
> min_bound = - max_bound
> bounds = (min_bound, max_bound)
> 
> optimizer = ps.single.GlobalBestPSO(n_particles=50, dimensions=ndim, options=options,
>                                     bounds=bounds,
>                                     ftol=1e-60,
>                                     ftol_iter=0
>                                     )
> cost, pos = optimizer.optimize(fx.sphere, iters=300, n_processes=None)
> 
> print(f"Position:{pos}, Cost:{cost}, Stopped at iteration: ", len(optimizer.cost_history))
> 

By setting `ftol_iter=0`, I need to disable it so I expect the cost be close to ZERO for the sphere() function.
But this the output **without my fix**:

```
pyswarms.single.global_best - INFO - Optimization finished | «ftol_iter=0» reached | 
Last iteration: 2 |
best cost: 0.10621578511912153, best pos: [ 0.08408654 -0.31487337]
```
It was stopped on 2nd iteration. Is this what it should be? **cost==0.106**?
### It's definitely wrong.
Just note that the this is the modified log by me. You don't see at what iteration the search has been stopped. In addition, **you won't see the next outputs with the original code because of the bug**

Now, output with my fix:
```
pyswarms.single.global_best - INFO - Optimization finished | «iters=300» reached | 
Last iteration: 300 |
best cost: 7.746188828069522e-16, best pos: [-2.04043748e-08  1.89282955e-08]
```
It reached to the max iteration (300), and the cost is 7.74e-16, close to ZERO.

Now, if I set `ftol=1e-3', a smaller value (ftol_iter is still disabled):
```
pyswarms.single.global_best - INFO - Optimization finished | «ftol=0.001» reached | 
Last iteration: 9 |
best cost: 0.007485035249975975, best pos: [-0.06714777 -0.05455467]
```
The search was stopped on 9th iteration, and the cost is 0.007. NOT BAD.


> This is the main purpose of `ftol` and if you don't want **two consecutive iterations**, you can specify `ftol_iter` to extend the property over more iterations.

As I said, if I disable ftol_iter, it isn't still working. There is no way to disable this behavior so I made that fix; but if someone wants to stop the search like that, one can set `ftol_iter=1'.
**ftol has been borrowed** from other software such as MATLAB. The idea of ftol doesn't work for PSO with the current implementation because of the global_best_cost.

Anyway, I haven't touched the ftol_iter implementation; just disabled it if `ftol_iter=0`, and changed the default value to **zero**. So it works as it has been designed for if `ftol_iter>=1`.

I hope that I explained everything clear and in details, for that last time by the way :smile: .

nishnash54 commented:
> When I set `ftol=10e-60`, I mean the gradient in changes of the best_cost must be less than `1e-60`, which is a very small value. It literally means never stop the search until max iter reaches.

So according to your implementation when there is no change in the `best_cost` over two iterations i.e. the change is exactly **0** which is less than `1e-60`, the algorithm must simply ignore this and keep running? The smallest possible value is **-inf** which is set as the default which **"literally"** makes sure the algorithm runs for max `iter`.

> As I said, if I disable ftol_iter, it isn't still working. There is no way to disable this behavior so I made that fix; but if someone wants to stop the search like that, one can set `ftol_iter=1'.

If you have noticed from the codebase, `ftol` plays no role in the algorithm other that providing a metric for early stopping. So the way to **"disable"** `ftol_iter` (you want the algorithm to run `iter` times without stopping) is either

1. Do not change `ftol`, leave it at its default value `-inf`
2. Change `ftol`, but set `ftol_iter = iter` so that the algorithm runs for all the iterations.

because setting `ftol_iter = 0` makes no logical sense as you are now comparing `ftol` across zero pairs of iterations.

> Anyway, I haven't touched the ftol_iter implementation; just disabled it if `ftol_iter=0`, and changed the default value to **zero**. So it works as it has been designed for if `ftol_iter>=1`.

In the review comments I have mentioned how the change have affected the behavior of `ftol_iter`, how the test fails and an additional test to try out.

Like before I again suggest that you can create a sample test bed that describes where the current implementation fails or how you would like to have some different results.

**Note:** If you are trying to implement `fmin` behavior in terms of `ftol` by setting `ftol_iter` to 0, I think we can have a separate PR to implement that without breaking the current implementation.

msat59 commented on pyswarms/single/local_best.py:
```diff
@@ -209,7 +218,9 @@ def optimize(self, objective_func, iters, n_processes=None, **kwargs):
-        for i in self.rep.pbar(iters, self.name):
+        # Default reason for optimization completion
+        stop_reason = "«iters={}» reached".format(iters)
+        for i in range(iters) if verbose else self.rep.pbar(iters, self.name):
```
Fixed. I forgot about the for loop.
I use the web UI. I must familiarize myself with git.

msat59 commented:
> If you have noticed from the codebase, ftol plays no role in the algorithm other that providing a metric for early stopping. So the way to "disable" ftol_iter (you want the algorithm to run iter times without stopping) is either
> 
> Do not change ftol, leave it at its default value -inf
> Change ftol, but set ftol_iter = iter so that the algorithm runs for all the iterations.
> because setting ftol_iter = 0 makes no logical sense as you are now comparing ftol across zero pairs of iterations.

@nishnash54 , Could you please just forget about ftol_iter implementation for one minute? I am not against your implementation; I don't want to remove your code.
I am talking about how ftol must work, without ftol_iter; when you haven't put it in the code, I mean before two weeks ago.

1-No one is going to play with ftol_iter at the beginning. So the algorithm **must work with its default settings** and definitely, without activating the ftol_iter. I mean when `ftol_iter` is disabled, the algorithm must converge. In the current implementation, **there is no way to disable ftol_iter** and improve the convergence. Setting ftol_iter=100 or 500 is a stupid solution (sorry man); even setting `ftol=-np.inf`. When you use abs(x1-x2), it's always >=0 so setting a negative value doesn't make sense. It's cheating to make the algorithm work. :wink: 

2-In the previous code provided, apply these changes: `ftol=1e-1`, `ftol_iter=1`. Again, it stops on the 3rd iteration because `swarm.best_cost - best_cost_yet_found = 0` (delta). Below is what I got for their difference (delta). It was much greater than 0.1 and it was increasing; but **suddenly**, it became zero !!!!
Why?
Because it couldn't find a better solution; not because it converged. 
```
Iter 1: 0.3045447712739723
Iter 2: 0.5754581630596829
Iter 3: 0.0
```

**There is no magical algorithm but a stupid one that thinks that it has converged on the 3rd iteration without being at the exact local/global optima**. You think it has converged because `swarm.best_cost == best_cost_yet_found`; but it is totallllllyyyyy wrrrrrooooonnnngggggg.

I don't know why you don't understand it.
Do you really know how PSO works? It's not a classical optimization algorithm.
If you know how it works, you should understand what I mean by saying  **In PSO, if the best cost doesn't change in two iterations, it does not indicate that the algorithm converges and we have the solution**.
PSO is not something like Newton Raphson method. If it was, you were right by saying no change in the cost so it converged; but if it was, no one needed the ftol_iter; it was pointless. As long as the delta was below ftol, the user was happy. 

### BUT PSO IS NOT A CLASSICAL MATHEMATICAL OPTIMIZATION METHOD.
In [Newton Raphson method](https://brilliant.org/wiki/newton-raphson-method/) for instance, the delta is getting smaller and smaller as we proceed; but it fluctuates in PSO; because it is a heuristic method.

The last word; pick any function or optimization problem, set any value for ftol (except _-np.inf_) then disable the ftol_iter and try to find the solution,  ZERO in case of the sphere() function. JUST DISABLE THE ftol_iter for one minute only.

ljvmiranda921 commented:
Hi @msat59 , I appreciate this Pull Request but please help keep our discussions here orderly and civil (and this goes to all of us involved in this PR). I understand where you're coming from, and sure the frustration is valid, but let's relax for a bit!

I know that everyone is donating their time and knowledge to solve this problem. I acknowledge you, @hakan-gecili and @nishnash54 's contributions on this matter. Let's take a step back and ensure that we get through this PR together :+1: 

For this I recommend the following:
* **Write down tests so that you can clearly define what your expected inputs and outputs are**. I think that this is the best way to talk through your point so that you don't need to repeat it over and over again. Show us cases like: (1) what happens if `ftol` is disabled, and `ftol_iter` is used? (2) what happens if `ftol_iter` is disabled, and `ftol` is used? (3) what happens if both are used? (4) what happens if both are disabled? I'm sure that once you've done this, everything will be much clearer. If you're not familiar with testing, @whzup and I can help you out. 
* **I'll be adding some comments regarding Python code style and formatting, I hope you don't mind**. We might also need to adopt some of our docstrings into a warm, friendly, and knowledgeable tone. I know it's warm-fuzzy, but I believe that it's a design decision that has helped this project be useful to beginners and advanced researchers alike.

Again, for everyone involved in this PR (of course including me). **Let's help keep our conversations healthy and civil.** This may be tough times in general (with all the things going around), and perhaps it may help if we can step back a little, but let's do an effort to keep this environment/PR healthy.

ljvmiranda921 commented on pyswarms/discrete/binary.py:
```diff
@@ -200,13 +216,22 @@ def optimize(self, objective_func, iters, n_processes=None, **kwargs):
+            costs_diff = np.abs(self.swarm.best_cost - best_cost_yet_found)
+            isBelow_ftol = costs_diff < relative_measure
+            
+            # Check if ftol reached, exclude no change in the best_cost
+            if isBelow_ftol and costs_diff != 0:
+                stop_reason = "«ftol={}» reached".format(self.ftol)
+                break
```
Hi @nishnash54 , thanks a lot for this. Some clarifications: I think we both agree on the idea of `ftol_iter`. Before this feature, `ftol` only checks the difference between two subsequent iterations. Now with `ftol_iter`, we check it for `ftol_iter` iterations. 

So right now in `master`, the default value of `ftol_iter` is `1`. Then this means that it checks if there are any improvement over `1` iteration. Am I correct to assume that this is like the "disable" switch for `ftol_iter`? or is this `0` like @msat59 's experiments. Just clarifying, although I feel that we're on the same page.

nishnash54 replied:
Yes, exactly. The default `ftol_iter` is set to `1`, which means `ftol_iter` is disabled and the initial behavior of `ftol` is maintained.

>Before this feature, ftol only checks the difference between two subsequent iterations.

Let's call "two subsequent iterations" a pair, then `ftol_iter` is set with the expectation, the `ftol` condition passes for `ftol_iter` consecutive pairs.

nishnash54 commented:
@ljvmiranda921 Thanks for moderating this conversation and providing your feedback on this PR.
The point you have mentioned about test cases is extremely important, it will enable us to tackle the problem quickly. I have also mentioned that a simple testbed (just a `.py` file) with a bunch of example problems will work as well.

@msat59 I am not defending "my" code in anyway. As a matter of fact, it's just an implementation of a Feature Request that I found would add value to the library (especially for research purposes). Just as I have helped you fix bugs and pass the test cases in your initial PR, I am only pointing out [some cases](https://github.com/ljvmiranda921/pyswarms/pull/401#discussion_r422044752) where your implementation fails. That is the point of open source; to contribute and improve the library.

>I am talking about how ftol must work, without ftol_iter; when you haven't put it in the code, I mean before two weeks ago.

The idea of `ftol` was implemented way back in this [FR](https://github.com/ljvmiranda921/pyswarms/issues/91#issuecomment-378977580). I don't think there have been any problems since then (@ljvmiranda921 do you recall anything) As `ftol`, is applied to iterations, setting `ftol_iter` to 0 (disable) would mean we are not applying `ftol`. This is from the view of an end user.

The way I would use `ftol` and `ftol_iter` parameters to optimize any objective function

1. Set only `ftol`, might have to reduce and try other values. 
2. If I think it still did not converge enough OR I don't know if it can improve anymore
3. Increase `ftol_iter`, this should give the algorithm some breathing space, a chance to improve over the next few iterations. As it's a sliding window approach (using `all()`), any improvement will reset the window (again require `ftol_iter` iterations of no improvements)

ljvmiranda921 commented:
>  The idea of ftol was implemented way back in this FR. I don't think there have been any problems since then (@ljvmiranda921 do you recall anything) As ftol, is applied to iterations, setting ftol_iter to 0 (disable) would mean we are not applying ftol. This is from the view of an end user.

Wow that was way back! I don't recall any problems with ftol from the last few versions. But sure, I think it was good that @msat59 is bringing these things up for discussion as well, good chance for us to revisit things. 

> 
>   1. Set only ftol, might have to reduce and try other values.
>   2. If I think it still did not converge enough OR I don't know if it can improve anymore
>   3. Increase ftol_iter, this should give the algorithm some breathing space, a chance to improve over > the next few iterations. As it's a sliding window approach (using all()), any improvement will reset the window (again require ftol_iter iterations of no improvements)

This is how I imagine the use-case would look like too. Now, applying this to @msat59 's sample test case:

```python
import numpy as np
import pyswarms as ps
from pyswarms.utils.functions import single_obj as fx

options = {'c1': 0.5, 'c2': 0.3, 'w':0.9}

ndim=2
max_bound = 5 * np.ones(ndim)
min_bound = - max_bound
bounds = (min_bound, max_bound)

optimizer = ps.single.GlobalBestPSO(n_particles=50, dimensions=ndim, options=options, bounds=bounds, ftol=1e-60, ftol_iter=0)
cost, pos = optimizer.optimize(fx.sphere, iters=300, n_processes=None)

print(f"Position:{pos}, Cost:{cost}, Stopped at iteration: ", len(optimizer.cost_history))
```

Then the best way to solve this would be to increase `ftol_iter` right? I tried setting it to the maximum number of iterations, `300`, and it does solve it. I know it's not what @msat59 is intending to do, but am I missing anything (or are there any potential repurcussions), to set or default `ftol_iter` to `n_iters`? By using this analogy, it gives the optimizer the "whole breathing room" since virtually all iterations are now being checked.

I'll try to be back in a few hours, I might not be able to reply to the next few messages!

nishnash54 commented:
> but am I missing anything (or are there any potential repurcussions), to set or default ftol_iter to n_iters? By using this analogy, it gives the optimizer the "whole breathing room" since virtually all iterations are now being checked.

This would be one of the ways to run the algorithm for all iterations as mentioned in the [comment](https://github.com/ljvmiranda921/pyswarms/pull/401#issuecomment-625741474). The repercussions of any of these two methods results in `ftol` not playing any part in the optimization loop meaning that there is no early stopping. It is exactly the same as default PSO without `ftol` and `ftol_iter`.

whzup commented:
Hello guys. I have some thoughts on this matter:

1) I think I know what @msat59 means by saying

> So the algorithm must work if the ftol is used. I mean when ftol_iter is disabled, the algorithm must converge.

when referring to `ftol_iter=0`. This may be a problem with the handling of an empty list and the `all()` function [here](https://github.com/ljvmiranda921/pyswarms/blob/master/pyswarms/single/general_optimizer.py#L264) (not 100% sure though). In my opinion, setting `ftol_iter=0` is a sensible way to not apply `ftol` at all *but* the ways mentioned by @nishnash54 are also valid. I think just one of them suffices and if the `ftol_iter=0` leads to a problem we can just mention this in the documentation and check for it in `__init__`.

 2) I would actually prefer to have `ftol_iter` as small as possible so we do not have to shift the array in every iteration if we do not use `ftol`. This is an expensive operation, [~O(ftol_iter-1)](https://wiki.python.org/moin/TimeComplexity) in our case, especially for big arrays. Maybe we should even consider using a [deque](https://docs.python.org/3.7/library/collections.html#collections.deque). Or we could just disable the `ftol_iter` functionality if `ftol` is not provided.

3) I think it might make sense to add `ftol_history` as an attribute to the optimizer, such that a user could inspect the delta behavior. It would also make this discussion easier as we could easily share the `ftol_history`.

nishnash54 commented:
@whzup Thanks for your time. You hit the bullseye.

>In my opinion, setting ftol_iter=0 is a sensible way to not apply `ftol` at all

This is exactly what I am trying to explain; that setting `ftol_iter = 0` would mean not applying `ftol` at all, which is the same as leaving `ftol` and `ftol_iter` at their default values. For some reason @msat59 wants to set `ftol` but disable `ftol_iter` (I don't see how that works). Maybe it's not clear in the documentation.

The part about setting `ftol_iter = 0` might throw errors, need to handle that in  `__init__` and also using `deque` should improve the efficiency. Also your idea about `ftol_history` is interesting.

I'll open a new PR with a few additional tests, updated documentation, example code and also try out if `deque` improves performance.

whzup commented:
> This is exactly what I am trying to explain; that setting ftol_iter = 0 would mean not applying ftol at all, which is the same as leaving ftol and ftol_iter at their default values. For some reason @msat59 wants to set ftol but disable ftol_iter (I don't see how that works).

I think his point is, that if you set `ftol_iter=0` the optimization ends prematurely. But this is not what you would expect when setting `ftol_iter=0`. Actually, consider the following example
```python
>>> all([])
True
```
Now if we have `ftol_iter=0` we have an empty list for `ftol_history` which means that the check
```python
if all(ftol_history):
    break
```
will default to `True`, thus breaking the loop (not the expected behavior). In this case I'd propose that we fix `ftol_iter > 0` and "turn off" the entire feature (whatever this will mean) if `ftol` is not provided.

msat59 commented on pyswarms/discrete/binary.py:
```diff
@@ -161,10 +168,16 @@ def optimize(self, objective_func, iters, n_processes=None, **kwargs):
         """
+        # Apply verbosity
+        if verbose:
+            logginglevel = logging.INFO
```
I made in like `logging.INFO`; but `log_level` seems fine too.

msat59 commented:
> > This is exactly what I am trying to explain; that setting ftol_iter = 0 would mean not applying ftol at all, which is the same as leaving ftol and ftol_iter at their default values. For some reason @msat59 wants to set ftol but disable ftol_iter (I don't see how that works).
> 
> I think his point is, that if you set `ftol_iter=0` the optimization ends prematurely. But this is not what you would expect when setting `ftol_iter=0`. Actually, consider the following example
> 
> ```python
> >>> all([])
> True
> ```
> 
> Now if we have `ftol_iter=0` we have an empty list for `ftol_history` which means that the check
> 
> ```python
> if all(ftol_history):
>     break
> ```
> 
> will default to `True`, thus breaking the loop (not the expected behavior). In this case I'd propose that we fix `ftol_iter > 0` and "turn off" the entire feature (whatever this will mean) if `ftol` is not provided.

It can be easily switched off with an `if`. It may not be the most efficient way; but it really doesn't matter. The number of iteration are usually few hundreds or thousands so it doesn't add anything to the total time. 

Regarding to your concern about `~O(ftol_iter-1)`, all(ftol_history) takes more time than nested `if` for bigger `ftol_iter`. For example, this is the approximate timing for `ftol_iter=50` for using all() (first line) and using nested if statements (second line); but you see it's nothing. However, deque may be a better solution programmatically.

```
8.55 µs ± 253 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
6.2 µs ± 179 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

msat59 commented:
@ljvmiranda921, did you find the reason for the conflicts?

msat59 commented:
> > In my opinion, setting ftol_iter=0 is a sensible way to not apply `ftol` at all
> 
> This is exactly what I am trying to explain; that setting `ftol_iter = 0` would mean not applying `ftol` at all, which is the same as leaving `ftol` and `ftol_iter` at their default values. For some reason @msat59 wants to set `ftol` but disable `ftol_iter` (I don't see how that works). Maybe it's not clear in the documentation.

@nishnash54, the reason I want to have `ftol `but not `ftol_iter `is to continue the search as long as the gradient of changes in the cost is above `ftol`. When it is below it (not being zero as I've explained why), stop the search.

If you guys, look at [this](http://www.sosmath.com/calculus/diff/der07/der07.html) page, it explains how Newton's Raphson method solves an optimization problem. You will never have zero gradient, however it may be very very tiny. 
![img17](https://user-images.githubusercontent.com/20813541/81870908-1ab5f280-9577-11ea-8f36-5c13e70005da.gif)


That's when `ftol` plays its role and why I want to have it (here we can assume the cost is f(x)/f'(x) ). In this example, the answer is X4 for `ftol=1e-3`; but it's X6 for `ftol=1e-19`.

As you see, the delta X is getting smaller and smaller and will be zero after infinite iterations. It doesn't become zero instantly, and then increases.
That's why I say: the story is different in metaheuristic methods like Particle Swarm Optimization, Genetic Algorithm, Ant Colony Optimization, Shuffled Frog Leaping Algorithm and so on.

These algorithm try to find the best solution (smallest cost) at i _th_ iteration among _N_ parallel solutions (n_particles). There is a high possibility of having NO PROGRESS, even at the 2nd iteration.

So you can see how it differs:
`delta==0 at 2nd iteration   **VS**   delta==0 at infinite iterations.`

That's all the reason behind skipping `np.abs(self.swarm.best_cost - best_cost_yet_found)==0`, disabling `ftol_iter` and having **only** 'ftol` to check the gradient.

As a conclusion, I rephrase my previous explanations in this way: 
**To make the behavior of metaheuristic algorithms like classic numerical optimization algorithms so to converge correctly and avoid incorrect early stopping is to ignore the delta cost==0**

whzup commented:
@msat59 Could you just clarify what you mean by "disabling `ftol_iter` and having only `ftol` to check the gradient"? Both `ftol` and `ftol_iter` are half of one feature and not separate features. You specify the acceptable tolerance of the differences by using `ftol` and then you specify how many differences you want to compare using `ftol_iter`. There is no using `ftol` without `ftol_iter`.

I see your point about the improvement over time and that it may prematurely break the optimization without actually even having started converging. But is it really a general problem if two consecutive costs are the same? As soon as `ftol_iter > 1` the probability of having `ftol_iter` consecutive costs be exactly the same decreases drastically. I agree that it is very possible for `ftol_iter=1` (as you showed above).  Now the other question is if the exclusion of no change in two consecutive costs may not lead to unexpected behavior in some cases.

msat59 commented:
> @msat59 Could you just clarify what you mean by "disabling `ftol_iter` and having only `ftol` to check the gradient"? Both `ftol` and `ftol_iter` are half of one feature and not separate features. You specify the acceptable tolerance of the differences by using `ftol` and then you specify how many differences you want to compare using `ftol_iter`. There is no using `ftol` without `ftol_iter`.
> 
> I see your point about the improvement over time and that it may prematurely break the optimization without actually even having started converging. But is it really a general problem if two consecutive costs are the same? As soon as `ftol_iter > 1` the probability of having `ftol_iter` consecutive costs be exactly the same decreases drastically. I agree that it is very possible for `ftol_iter=1` (as you showed above). Now the other question is if the exclusion of no change in two consecutive costs may not lead to unexpected behavior in some cases.

@whzup , Is my English that bad that I can't clarify the reason? :wink: 

> There is no using `ftol` without `ftol_iter`

Of course there is. If you've ever used MATLAB and _fminsearch_ to solve your adv. math. course problems, you know what are `TolFun` and `TolX` and you don't have anything like ftol_iter. Early stopping is a good feature to have; but not always.

There are different scenarios:

1A- As an Engineer, I have several cases that I need the best possible solution; I don't care about the time. I need to optimize my chemical plants for instance to have minumum cost or maximum production, so I don't care if reaching below `ftol=1e-6` takes one minute or one week. I don't care about the time because spending one week (may cost $2000 for the company), which is much cheaper than solving the solution in 1 hour, but the cost of production becomes $2000 more everyday.
As you see, I want to bring down the objective function cost to a certain limit so I **don't need early stopping** (`ftol_iter`). I ONLY WANT TO KNOW WHAT X VALUES  GIVE ME **delta Y <=** `ftol`.

1B- For a different problem, I may need to have a quick evaluation; so I need to stop the search if 
a) deltaY<= 'ftol' 
or
b) I don't see any progress in changes of deltaY, after 10 iterations for instance (`ftol_iter=10`)

Because the `ftol` may be too small, or PSO cannot find the global optimum, or I want a quick solution/approximation so time is more valuable for me that the best solution. In this case, I NEED **BOTH** `ftol` and `ftol_iter`

2- As a data scientist, I define the objective function as the changes in the accuracy, precision, R2, or MSE of the _validaion set_ so I may not need to specify any exact value for `ftol` (I may set `1e-10` but definitely don't set `-np.inf` to disable early stopping) but I want to stop the search if there is no progress after 5 iterations (`ftol_iter=5`); because the _training set_ is being overfitted if I proceed so I need to use the early stopping feature and stop after 5 iterations with no change.

Is it clear enough buddy?

People may have experienced all three scenarios: 
with `ftol` and without `ftol_iter`  (best/close to exact solution, time doesn't matter)
with `ftol` and with `ftol_iter`    (quick approximation, time matters)
without (any specific) `ftol` (but not -np.inf) and with `ftol_iter`  (avoid overfitting)

So the algorithm must, I repeat, MUST work for all cases.

> As soon as ftol_iter > 1 the probability of having ftol_iter consecutive costs be exactly the same decreases drastically.

Not necessarily. It depends on n_particles, particle position, velocity, the nature of the problem and so on. For the provided sample (sphere), the search can be stopped on 14 th iteration, 10 iterations without improvement for `ftol=1e-10`. Look again: **10 iterations without improvement** , without even getting close to `1e-10`. It jumped from `0.02` to `0`.
This is the delta best_cost (costs_diff in my fix) I got:

> inf
> 0.0
> 0.5436045970066082
> 0.020231498859684414
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> Position:[0.0032389], Cost:1.0490498434762185e-05, Stopped at iteration:  14

All of us know clearly that the position must be close to zero.
By disabling the early stopping, `ftol_iter=0` in my fix, and `ftol=1e-10` you can get
 
`Position:[-2.16444468e-06], Cost:4.684820760526288e-12, Stopped at iteration:  176`

and below is the final sequence of costs_diff by disabling the early stopping, `ftol_iter=0`

> 0.0
> 0.0
> 2.9170793487114684e-10
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 0.0
> 2020-05-14 12:51:40,727 - pyswarms.single.global_best - INFO - Optimization finished | «ftol=1e-10» reached | Last iteration: 176) |
> best cost: 4.684820760526288e-12, best pos: [-2.16444468e-06]

After 17 iterations with no change, the `costs_diff` is below `ftol`, finally; and the X is so much closer to zero, 2.16e-6).

I hope to be clear enough. 

Guys, it is a simple logic and tweak to fix the bug; but I have explained it several times to everybody, spending more than 30 minutes each time. I am scared to introduce new features such as velocity damping, or adaptive PSO then you force me to write a book or lecture about it a semester.
(for this example, I had costs_diff<1e-10 after 176 iteration, but can get same result after 30 iterations by implementing the **damping**)

@ljvmiranda921 , I have covered everything so I don't see any need to explain things again and again.
please resolve the conflict issues then I can finalize the changes.

The last word: 
My fix ignores costs_diff=0 if `ftol_iter=0` (disabled); but if anyone wants to have it, should set `ftol_iter>=1`. As a result, my fix **DOES NOT CHANGE** the functionality of the `ftol_iter`, AKA early stopping.
### SO DON'T GET SCARED  :wink:

whzup commented:
@msat59 Well, I think we are talking about different things here and you obviously did not read my last comment above thoroughly. I just looked at the changes and saw that you changed how `ftol` and `ftol_iter` work. I think `ftol_iter` was meant as a generalization of `ftol` in the sense that the function of `ftol` is included as a special case when `ftol_iter = 1`. As far as I'm concerned it is not sensible to implement the special case separately and disable `ftol_iter` for that. I'd argue that what you say is the behavior when "`ftol` is enabled and `ftol_iter` is disabled" should be the behavior of `ftol_iter = 1`. In my opinion, `ftol` should **only** be responsible for the maximally acceptable difference (not for how many differences are compared) and `ftol_iter` is **only** responsible for how many differences are compared. This clearly distinguishes the two. No `ftol` without `ftol_iter` and vice versa.

Also, you having to explain 30 minutes every time is a consequence of you not presenting your information and arguments clearly!
1. You did not try to understand what we wanted to say, we pointed out multiple times, that in our understanding `ftol` and `ftol_iter` are linked in a way such that you cannot disable one without disabling the other.
2. You did not investigate your findings thoroughly. In your [first comment here](https://github.com/ljvmiranda921/pyswarms/pull/401#issuecomment-625537931) you even failed to spot the `all([])` bug and blamed it on a bad implementation of `ftol_iter`. You also did not provide any code.
3. It was also very hard to distill what exactly you believed `ftol` and `ftol_iter` should do (you only pointed out what you thought they did wrongly). @nishnash54 asked you to specifically explain what you expect to happen and what the features should do, which you did not do.
4. You went on unnecessary tangents about how you have worked with GA for multiple years and the scenarios which you just mentioned in the last comment. Not to mention the very disrespectful phrases sprinkled in.

I expect you to answer with concise arguments and **explicit** examples from now on.

msat59 commented:
> 1. You did not try to understand what we wanted to say, we pointed out multiple times, that in our understanding `ftol` and `ftol_iter` are linked in a way such that you cannot disable one without disabling the other.

Dear @whzup , I understand what you and @nishnash54  are trying to say; but your understanding is not correct.  They don't have to be linked necessarily otherwise you intentionally want to have it like that and then have an incomplete implementation. As I have shown in my fix, they can work separately and together.

> 2. You did not investigate your findings thoroughly. In your [first comment here](https://github.com/ljvmiranda921/pyswarms/pull/401#issuecomment-625537931) you even failed to spot the `all([])` bug and blamed it on a bad implementation of `ftol_iter`. You also did not provide any code.
It's not my implementation so @nishnash54  is the right person to explain it. It's a better implementation than mine (nested if statements) but it can be fixed without touching the current form.

> 3. It was also very hard to distill what exactly you believed `ftol` and `ftol_iter` should do (you only pointed out what you thought they did wrongly). @nishnash54 asked you to specifically explain what you expect to happen and what the features should do, which you did not do.

I explained more than 10 times. No need to mention again. Please read my previous posts that how the algorithm should work with/without setting `ftol` and `ftol_iter`.

> 4. You went on unnecessary tangents about how you have worked with GA for multiple years and the scenarios which you just mentioned in the last comment. Not to mention the very disrespectful phrases sprinkled in.
> 

I know that it was irrelevant, but it seemed that nobody really understood the real meaning of `ftol` and everybody seemed to talk based on their thoughts, which may be right or wrong.  I should explain that I was talking based on 13 years of experiences in data analysis and modeling and 20 years of being an Engineer. I'm not a fresh graduate that recently has learn how to code or solve a Sphere() function.

> I expect you to answer with concise arguments and **explicit** examples from now on.

I've already given the code and the expected behavior of the algorithm; but if you need a real problem beside that simple Sphere() function I should spend more time to find different case studies.
Do you really think that I should prove that the algorithm is not working properly again?
If the answer is YES, I try to submit different case studies!

msat59 commented:
Guys,

If `ftol_iter=0` doesn't make sense to you, the default value can be set to 'np.inf` too. I think it's more understandable to everyone. The only issue is using np.array to count the iterations without improvement ( ftol_history = [None] * self.ftol_iter). This can be resolved by  self.ftol_iter = iters inside the optimze() method.

@ljvmiranda921 
@whzup 
@nishnash54 
What is your idea about this fix?

The time that I was trying to solve the premature solution, @nishnash54 hadn't implement the `ftol_iter` yet so I had to ignore delta cost=0. Now, if we set the default value to maximum iterations, there is no need to work around delta cost=0

msat59 commented:
I was thinking about setting default value to `ftol_iter=np.inf` ; but it doesn't work in case of needing `ftol` only.

nishnash54 commented:
@whzup @ljvmiranda921  and I have been trying to explain our understanding of the implementation, how it was developed and how it is backward compatible with the initial implementation of `ftol`.

We have requested multiple times to upload a sample test bed (a few examples) in code so that we can compare with the current implementation and your updated algorithm. If you do have time I suggest you provide some additional case studies as it would help a lot.

As you keep mentioning that your changes do not break anything and maintain the functionality of `ftol_iter`, let me point you back to the [comment](https://github.com/ljvmiranda921/pyswarms/pull/401#discussion_r421749479). Also your code does not pass the unit tests (this is the part that is worrying as it proves something is broken) and the additional example I have provided. If this behavior is expected from the implementation, I do not see how `ftol_iter` works with early stopping in your implementation. Please do look into this.

msat59 commented:
I was looking for an old PSO problems on my computer then I found an old PSO toolbox for MATLAB, when MATLAB hadn't implemented _particleswarm()_ yet.

It uses a very interesting way to exclude **iterations with no progression** from costs_diff<ftol comparison; but also keep the track of them to satisfy `ftol_iter` (MaxStallIterations in MATLAB).

I should implement it and check the results. I also need to install MATLAB and compare the results. I will do it during this weekend.

I therefore need to undo my changes and implement the new one, as there is a conflict between my commits and the master.

Please advise me how to undo the changes; but no closing the PR. Should I remove the commits from my fork?
If it wor

ljvmiranda921 commented:
> I should implement it and check the results. I also need to install MATLAB and compare the results. I will do it during this weekend.

Thanks for this @msat59 . To avoid running circles around this topic, we'll wait for your implementation of the one above. I think it's easier to communicate our ideas if we do a simple test bed, as @nishnash54 suggests. 

Don't worry about Git branches fow now. Once we've settled on the proper business logic, then the technical implementation will be more straightforward :+1:

msat59 commented:
Hi Guys.

Before testing the code, let me explain something.

The logic behind the ftol_iter and previous discussions is clearer if we rename `ftol_iter` to `MaxStallIter` or `NoChangIter` for now (just to clarify the idea, however, I think `ftol_iter` name is a bit confusing and something like `stall_iter` is more intuitive). I will explain later.


The logic is:

1- If a better/smaller `best_cost`  is found, then check if best_cost[i] - best_cost[i-1] < relative_ftol. In other word, `np.abs(self.swarm.best_cost - best_cost_yet_found) < relative_measure`. If it is satisfied, then **stop the search**.
In this fashion, we don't have premature solution because of **no progression**.

2- If `best_cost` doesn't change, then check if the number of iterations without change is <= `nochange_iter`/`stall_iter`/`ftol_iter`, then continue otherwise **stop the search.**
So if the algorithm cannot find any better solution after a defined number of iterations, the **early stop** comes into effect.

OK, as you see, `ftol_iter` actually defines the number of iterations without improvement; but when I started to play with it last week, I thought it was the number of iterations if `delta cost < ftol` .  That was why I was trying to say that it did not make sense.
In all optimization algorithms, the search is stopped when detla cost < ftol. In addition, when we define early stopping and we say stop the search if the change in cost is less than 0.1 and if there is no progress after 10 iters, we **do not mean** if the change is less than 0.1 for 10 iters. **DO WE?**
If the search takes a long time and we want to avoid pointless long calculations, why should we continue searching for 9 more iterations if the change is below 0.1? Once the condition is met, we should stop because we are satisfied with the result (below 0.1).

As a result, I believe that `ftol_iter` is confusing and it should be changed to something like `stall_iter` or `nochange_iter`.

Now, we have two options for the default value of `ftol_iter` or whatever we name it.

**Option A:** Set the default value to _np.inf_. In this implementation, _np.array_ has not been used so _np.inf_ can be used without any problem to **DISABLE NO PROGRESSION CHECK** if the user doesn't need it. It is what @nishnash54  and @whzup had concern about the functionality of `ftol_iter` and we argued a lot about it. In addition, `ftol_iter` MUST be GREATER THAN ZERO, and @whzup  mentioned that  `ftol_iter=0` didn't make sense (but I think _np.inf_ makes sense now)

**Option B:** Set the default value to something big enough to avoid unwanted early stopping. It is 20 in MATLAB and it was 50 in that old toolbox. I think 50 is better if you are not happy with **option A**.

OK. I will commit the changes after everybody were happy about this implementation. I've only modified the `global_best.py` so replace it with yours to check its code

.
**Note 1:** I've added few prints to log the changes in `i, swarm.best_cost, best_cost_yet_found, their difference, and relative_measure` so the `verbose=False` is used to avoid the conflict in the progress bar. 
**Note 2:** It's not the final release, some variables name may be changed. Current names are for more clarifications only.
**Note 3:** I have also used that toolbox code to implement the `fmin` or `flim` (fitness limit) check, however I've disabled it by `FitnessLimit = -np.inf`; so will be implemented in another PR later.


### Let's check the code. 

There are three possible scenarios in the archive attached. It's a multi-constraint optimization problem with three variables (3D search space). The minimum cost I found after 10 reps was `Cost: 9.394055823182274` and `X ≈ [0.43807206 1.45667361 3.10525434]`

[Sample 1.zip](https://github.com/ljvmiranda921/pyswarms/files/4641310/Sample.1.zip)

1- `PSO_1_MaxIters.py` : Uses nIters (MaxIter) only, so `ftol=-np.inf`. So there is no early stopping condition. The algorithm does not care if `best_cost` remains constant.
Typical plot (yours is different):
![PSO_1_MaxIters](https://user-images.githubusercontent.com/20813541/82163017-b2358100-98a8-11ea-89e6-ac43779aeb61.png)

2- `PSO_1_ftol_only.py` : Uses ONLY `ftol=1e-10`. In my fix, I disabled/ignored `ftol_iter` by setting `ftol_iter=-np.inf`. So the algorithm stops at the **first** occurrence of delta best_cost < ftol. It doesn't care if best_cost remains constant (costs_diff=0) as I have mentioned the reason previously, which results in an immature solution and unwanted early stopping.
Typical plot:
![PSO_1_ftol_only](https://user-images.githubusercontent.com/20813541/82163031-c7121480-98a8-11ea-9802-cf5e479e0891.png)

Some flat areas are seen, but they are ignored because of `ftol_iter=np.inf`.

3- `PSO_1_ftol_ftoliter.py` : uses `ftol=1e-20` and `ftol_iter=10`. The algorithm stops the search if costs_diff<1e-20, or no progress observed for 10 iterations.
Typical plot:
![PSO_1_ftol_ftoliter](https://user-images.githubusercontent.com/20813541/82163156-aa2a1100-98a9-11ea-930e-28316ea2c78b.png)

At the right end, there is a flat line for 10 iterations because having no progress in reducing the best cost. 

I hope that I explained everything clear.

### Please give your feedbacks.

msat59 commented:
Isn't there any feedback after three weeks? :confused:

ljvmiranda921 commented:
Oops, sorry for this. The notif got lost and I thought I'm just waiting for your reply. Let me check this weekend. Initial thoughts on your end, @whzup ?

whzup commented:
@ljvmiranda921 I have to look into this in more details, I still feel like there is a big misunderstanding in what the purpose of this feature is.
